[SPARK-11077] [SQL] Join elimination in Catalyst

ankurdave opened the pull request:
Join elimination is a query optimization where certain joins can be eliminated when followed by projections that only keep columns from one side of the join, and when certain columns are known to be unique or foreign keys. This can be very useful for queries involving views and machine-generated queries.

This PR adds join elimination by (1) supporting unique and foreign key hints in logical plans, (2) adding methods in the DataFrame API to let users provide these hints, and (3) adding an optimizer rule that eliminates unique key outer joins and referential integrity joins when followed by an appropriate projection.

This change is described in detail here: https://docs.google.com/document/d/1-YgQSQywHfAo4PhAT-zOOkFZtVcju99h3dYQq-i9GWQ/edit?usp=sharing

ankurdave committed:
Eliminate outer join before project

ankurdave committed:
Use KeyHint to do join elimination

ankurdave committed:
Add foreign keys

ankurdave committed:
Alias-aware join elimination + bugfixes

ankurdave committed:
Propagate foreign keys through Join operator

ankurdave committed:
Remove key hints after join elimination

ankurdave committed:
Support inner joins based on referential integrity

ankurdave committed:
Correctness fixes for join elimination

Do not eliminate referential integrity full outer joins, or inner joins where foreign key is
nullable. Require foreign keys to reference unique columns.

ankurdave committed:
Do key hint resolution during analysis

This is necessary to support aliased self joins and multiple foreign keys with the same referent.

ankurdave committed:
Don't crash when foreign key refers to unresolved relation

Instead just leave the KeyHint unresolved.

ankurdave committed:
Fix JoinEliminationSuite

ankurdave committed:
Merge remote-tracking branch 'apache-spark/master' into GraphFrames

ankurdave committed:
Fix KeyHintSuite after merge

ankurdave committed:
In ForeignKey, store referencedRelation as logical plan

Previously we stored its name as part of referencedAttr, requiring a
catalog lookup.

ankurdave committed:
Use semanticEquals for Attributes

ankurdave committed:
Remove TODOs

ankurdave committed:
Add more comments

ankurdave committed:
Merge remote-tracking branch 'apache-spark/master' into GraphFrames

ankurdave committed:
Use SharedSQLContext in KeyHintSuite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43619/consoleFull) for   PR 9089 at commit [`578797c`](https://github.com/apache/spark/commit/578797c456e20d0fb07bf10cb3e64f09065948f9).

SparkQA commented:
[Test build #43619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43619/console) for   PR 9089 at commit [`578797c`](https://github.com/apache/spark/commit/578797c456e20d0fb07bf10cb3e64f09065948f9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43619/
Test FAILed.

ankurdave committed:
Remove long URLs

They were references to the join elimination logic in Teradata, which is
really just a standard optimization rule.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43620 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43620/consoleFull) for   PR 9089 at commit [`7c7357b`](https://github.com/apache/spark/commit/7c7357bf9c1e8bab3f2d828dd8bc3d6f7d851196).

SparkQA commented:
[Test build #43620 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43620/console) for   PR 9089 at commit [`7c7357b`](https://github.com/apache/spark/commit/7c7357bf9c1e8bab3f2d828dd8bc3d6f7d851196).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43620/
Test PASSed.

ankurdave commented:
@marmbrus I addressed your comments from the review about a month ago:

1. Foreign key references now store the referenced relation directly as a logical plan rather than requiring a catalog lookup.
2. We now use `semanticEquals` and `AttributeSet` for attributes instead of normal equality.

There were a few comments that didn't make sense on second thought:

1. Move the attribute equivalence check in `ForeignKeyFinder` to a method on `LogicalPlan`. We thought this would simplify the logic, but it turned out not to (still need to maintain the disjoint-set data structure, and the logic gets split between `LogicalPlan` and `Project`).
2. Move foreign key attribute resolution to its own rule that runs at the end of analysis. This would work fine, but it seems to fit well within `ResolveReferences`.

Finally, the new DataFrame methods should probably be marked as alpha somehow, but I'm not sure of the best way. Maybe a new ScalaDoc group?

cc @rxin, @jkbradley

rxin commented:
We can tag them as Experimental (even though the entire DataFrame API is experimental!)

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -362,14 +362,35 @@ class Analyzer(
+                          attributeRewrites.get(referencedAttr).getOrElse(referencedAttr))
+                      case other => other
+                    }
+                    KeyHint((keys ++ newKeys).distinct, child)
```
Can't we just use `newKeys` here? Why do we need to keep old keys?

ankurdave replied:
Good eye! This is to accommodate future self-joins. If we got rid of the old foreign keys, a future self-join would not recognize that the new keys applied to it, because the attributes would have been rewritten.  I just added a comment noting this.

There's [a unit test](https://github.com/apache/spark/pull/9089/files#diff-09ca3beb9c48d89b5fcf248e48d888ddR261) that covers this (fails if you remove the old keys).

ankurdave committed:
Fix override of KeyHint#transformExpressions{Up,Down}

ankurdave committed:
Declare new DataFrame methods extra-experimental

ankurdave committed:
Explain why we keep old keys in self-join rewrite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

ankurdave commented:
@rxin Thanks, I added the Experimental tags.

SparkQA commented:
[Test build #43633 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43633/consoleFull) for   PR 9089 at commit [`55bb135`](https://github.com/apache/spark/commit/55bb1354efcef98944caf96f8d59dc2f4a6459c0).

SparkQA commented:
[Test build #43633 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43633/console) for   PR 9089 at commit [`55bb135`](https://github.com/apache/spark/commit/55bb1354efcef98944caf96f8d59dc2f4a6459c0).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43633/
Test FAILed.

ankurdave commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43638/consoleFull) for   PR 9089 at commit [`55bb135`](https://github.com/apache/spark/commit/55bb1354efcef98944caf96f8d59dc2f4a6459c0).

SparkQA commented:
[Test build #43638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43638/console) for   PR 9089 at commit [`55bb135`](https://github.com/apache/spark/commit/55bb1354efcef98944caf96f8d59dc2f4a6459c0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43638/
Test PASSed.

jkbradley commented:
Calling uniqueKey on a DataFrame throws out the column names.  Is that intended?

ankurdave committed:
Revert "Fix override of KeyHint#transformExpressions{Up,Down}"

This reverts commit 50717599f1eb5bf2184a6b1df2e0aebabdebddec.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

ankurdave commented:
@jkbradley Oops, thanks for catching that. I introduced it in 50717599f1eb5bf2184a6b1df2e0aebabdebddec because I misunderstood the function of `transformExpressionsDown`. Should be fixed now.

SparkQA commented:
[Test build #43757 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43757/consoleFull) for   PR 9089 at commit [`e1ec23d`](https://github.com/apache/spark/commit/e1ec23da83d02adafbe1fdc7852e258f9289d293).

ankurdave committed:
Update transformExpressions override comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43758/consoleFull) for   PR 9089 at commit [`0cd8a91`](https://github.com/apache/spark/commit/0cd8a9185e96d0f21a8bd9a437c124566b9f2ce1).

SparkQA commented:
[Test build #43757 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43757/console) for   PR 9089 at commit [`e1ec23d`](https://github.com/apache/spark/commit/e1ec23da83d02adafbe1fdc7852e258f9289d293).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43757/
Test PASSed.

SparkQA commented:
[Test build #43758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43758/console) for   PR 9089 at commit [`0cd8a91`](https://github.com/apache/spark/commit/0cd8a9185e96d0f21a8bd9a437c124566b9f2ce1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `
  * `sealed abstract class Key `
  * `case class UniqueKey(attr: Attribute) extends Key `
  * `case class ForeignKey(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43758/
Test PASSed.

jkbradley commented:
@ankurdave  Np, thanks for the fix.  Btw, should the fix be accompanied by a unit test to catch that issue?

ankurdave committed:
Merge remote-tracking branch 'apache-spark/master' into SPARK-11077-JoinElimination

AmplabJenkins commented:
Build triggered. sha1 is merged.

AmplabJenkins commented:
Build started sha1 is merged.

SparkQA commented:
**[Test build #45232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45232/consoleFull)** for PR 9089 at commit [`5abceae`](https://github.com/apache/spark/commit/5abceaebadedc130feeab7aec8b97f4fac3bdfda).

SparkQA commented:
**[Test build #45232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45232/consoleFull)** for PR 9089 at commit [`5abceae`](https://github.com/apache/spark/commit/5abceaebadedc130feeab7aec8b97f4fac3bdfda).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class KeyHint(newKeys: Seq[Key], child: LogicalPlan) extends UnaryNode `\n  * `sealed abstract class Key `\n  * `case class UniqueKey(attr: Attribute) extends Key `\n  * `case class ForeignKey(`\n

AmplabJenkins commented:
Build finished. 5912 tests run, 0 skipped, 0 failed.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45232/

rxin commented:
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one.

jzhuge commented:
Is there any desire to resurrect this PR? e.g., without any breaking changes (new hint only, no api change)
